tests/dataloader/test_metric.py

```diff
@@ -0,0 +1,585 @@
+			for i in range(len(input['reference_key'][turn])):
+				max_length = input['reference_len_key'][turn][i]
+
+				length_sum += max_length - 1
+				for j in range(max_length - 1):
+					# word_loss += -(input['gen_prob_key'][i][j][ FakeDataLoader().vocab_to_index[input['reference_key'][i][j + 1]] ])
+					word_loss += -(input['gen_prob_key'][turn][i][j][ input['reference_key'][turn][i][j + 1] ])
+		return np.exp(word_loss / length_sum)
+
+	def test_close1(self):
+		dataloader = FakeDataLoader()
+		data = dataloader.get_data('reference_key', 'reference_len_key', 'gen_prob_key',\
+								   'gen_key', 'post_key', 'resp_key', 'res_key', \
+								   True, False, True)
+
+		ans = self.get_perplexity(data)
+
+		pm = MultiTurnPerplexityMetric(dataloader, 'reference_key', 'reference_len_key', 'gen_prob_key', full_check=True)
+		pm.forward(data)
+		res = np.isclose(pm.close()['perplexity'], ans)
+		assert res.all()
+
+	def test_close2(self):
+		dataloader = FakeDataLoader()
+		data = dataloader.get_data('reference_key', 'reference_len_key', 'gen_prob_key',\
+								   'gen_key', 'post_key', 'resp_key', 'res_key', \
+								   True, False, True)
+
+		print(data['reference_key'].shape)
+		data['reference_key'] = np.delete(data['reference_key'], 1)
+		print(data['reference_key'].shape)
+
+		pm = MultiTurnPerplexityMetric(dataloader, 'reference_key', 'reference_len_key', 'gen_prob_key', full_check=True)
+
+		try:
+			pm.forward(data)
+		except ValueError:
+			pass
+		else:
+			raise Exception('Error')
+
+	def test_close3(self):
+		dataloader = FakeDataLoader()
+		data = dataloader.get_data('reference_key', 'reference_len_key', 'gen_prob_key',\
+								   'gen_key', 'post_key', 'resp_key', 'res_key', \
+								   False, True, True)
+		ans = self.get_perplexity(data)
+
+		pm = MultiTurnPerplexityMetric(dataloader, 'reference_key', 'reference_len_key', 'gen_prob_key', full_check=True)
+		try:
+			pm.forward(data)
+		except ValueError:
+			pass
+		else:
+			raise Exception('Error')
+
+	def test_close4(self):
+		dataloader = FakeDataLoader()
+		data = dataloader.get_data('reference_key', 'reference_len_key', 'gen_prob_key',\
+								   'gen_key', 'post_key', 'resp_key', 'res_key', \
+								   True, True, True)
+		ans = self.get_perplexity(data)
+
+		pm = MultiTurnPerplexityMetric(dataloader, 'reference_key', 'reference_len_key', 'gen_prob_key', full_check=True)
+		pm.forward(data)
+		res = np.isclose(pm.close()['perplexity'], ans)
+		assert res.all()
+
+class TestBleuCorpusMetric:
+	def get_bleu(self, dataloader, input):
+		bleu_value = 0
+		sentence_num = 0
+		for gen_sen, resp_sen in zip(input['reference_key'], input['gen_key']):
+			gen_sen_processed = dataloader.trim_index(gen_sen)
+			resp_sen_processed = dataloader.trim_index(resp_sen)
+			bleu_value += sentence_bleu([resp_sen_processed[1 : ]], gen_sen_processed, \
+										smoothing_function=SmoothingFunction().method7)
+			sentence_num += 1
+		return bleu_value / sentence_num
```

Comment:
corpus_blue is not always equals to the average of sentence_bleu. (Although it is right when there is only one sentence.) Refer to: http://www.nltk.org/_modules/nltk/translate/bleu_score.html
I think it may not necessary to check the bleu value of a random string, because you must implement the ground truth in the same way as the lib. I think that you may find some examples (include some corner cases), and give a predefined answer. For example, when gen == resp, the bleu_value must be 1

Reply:
But if 'metric.py' uses corpus_bleu function from ntlk, I think whatever corner cases won't make it fail. Or if it fails, the question will be in ntlk.